### ISSUE TITLE
GPU g5.4xlarge, training pythia-2.8b model exits with return code = -9

### ISSUE THREAD

**Issue by ipfloater** [databrickslabs/dolly#194]
I have searched all the existing issues. Using latest dolly code as of June 6, 2023. The Readme.md says can train on 6.9B param on A10 instances (g5.24xlarge).

However I can't get hold of **g5.24xlarge**, instead run on **g5.4xlarge**.  Is this not enough?

System: g5.4xlarge, 64GB Memory, 1GPU;
DBR: 12.2 LTS

Set the following in widgets:
gpu_family="a10"
"num_gpus"=1
input_model="EleutherAI/pythia-2.8b"

Here is the log:

[2023-06-07 02:38:13,229] [WARNING] [runner.py:186:fetch_hostfile] Unable to find hostfile, will proceed with training with local resources only.
[2023-06-07 02:38:13,236] [INFO] [runner.py:550:main] cmd = /local_disk0/.ephemeral_nfs/envs/pythonEnv-615f6812-9c5d-4630-9be1-4a471fc90e4b/bin/python -u -m deepspeed.launcher.launch --world_info=eyJsb2NhbGhvc3QiOiBbMF19 --master_addr=127.0.0.1 --master_port=29500 --module --enable_each_rank_log=None training.trainer --input-model EleutherAI/pythia-2.8b --deepspeed /Workspace/Repos/xxxxxx/dolly/config/a10_config.json --epochs 2 --local-output-dir /local_disk0/dolly_training/dolly__2023-06-07T02:38:06 --dbfs-output-dir /dbfs/dolly_training/dolly__2023-06-07T02:38:06 --per-device-train-batch-size 4 --per-device-eval-batch-size 4 --logging-steps 10 --save-steps 200 --save-total-limit 20 --eval-steps 50 --warmup-steps 50 --test-size 200 --lr 5e-6 --bf16 true
[2023-06-07 02:38:16,810] [INFO] [launch.py:142:main] WORLD INFO DICT: {'localhost': [0]}
[2023-06-07 02:38:16,810] [INFO] [launch.py:148:main] nnodes=1, num_local_procs=1, node_rank=0
[2023-06-07 02:38:16,810] [INFO] [launch.py:161:main] global_rank_mapping=defaultdict(<class 'list'>, {'localhost': [0]})
[2023-06-07 02:38:16,810] [INFO] [launch.py:162:main] dist_world_size=1
[2023-06-07 02:38:16,810] [INFO] [launch.py:164:main] Setting CUDA_VISIBLE_DEVICES=0
2023-06-07 02:38:19.283879: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-06-07 02:38:28 INFO [__main__] Loading tokenizer for EleutherAI/pythia-2.8b
2023-06-07 02:38:28 INFO [__main__] Loading model for EleutherAI/pythia-2.8b
2023-06-07 02:39:38 INFO [__main__] Found max lenth: 2048
2023-06-07 02:39:38 INFO [__main__] Loading dataset from databricks/databricks-dolly-15k
2023-06-07 02:39:47 WARNING [datasets.builder] Found cached dataset json (/root/.cache/huggingface/datasets/databricks___json/databricks--databricks-dolly-15k-6e0f9ea7eaa0ee08/0.0.0/e347ab1c932092252e717ff3f949105a4dd28b27e842dd53157d2f72e276c2e4)
100%|█████████████████████████████████████████████| 1/1 [00:00<00:00, 41.23it/s]
2023-06-07 02:39:47 INFO [__main__] Found 15011 rows
2023-06-07 02:39:48 INFO [__main__] Preprocessing dataset                       
2023-06-07 02:39:56 INFO [__main__] Processed dataset has 15011 rows            
2023-06-07 02:39:59 INFO [__main__] Processed dataset has 14977 rows after filtering for truncated records
2023-06-07 02:39:59 INFO [__main__] Shuffling dataset
2023-06-07 02:39:59 INFO [__main__] Done preprocessing
2023-06-07 02:39:59 INFO [__main__] Train data size: 14777
2023-06-07 02:39:59 INFO [__main__] Test data size: 200
[2023-06-07 02:39:59,322] [INFO] [comm.py:652:init_distributed] Initializing TorchBackend in DeepSpeed with backend nccl
2023-06-07 02:39:59 INFO [torch.distributed.distributed_c10d] Added key: store_based_barrier_key:1 to store for rank: 0
2023-06-07 02:39:59 INFO [torch.distributed.distributed_c10d] Rank 0: Completed store-based barrier for key:store_based_barrier_key:1 with 1 nodes.
2023-06-07 02:39:59 INFO [__main__] Instantiating Trainer
2023-06-07 02:39:59 INFO [__main__] Training
2023-06-07 02:40:03 INFO [torch.distributed.distributed_c10d] Added key: store_based_barrier_key:2 to store for rank: 0
2023-06-07 02:40:03 INFO [torch.distributed.distributed_c10d] Rank 0: Completed store-based barrier for key:store_based_barrier_key:2 with 1 nodes.
Installed CUDA version 11.3 does not match the version torch was compiled with 11.7 but since the APIs are compatible, accepting this combination
Using /root/.cache/torch_extensions/py39_cu117 as PyTorch extensions root...
Creating extension directory /root/.cache/torch_extensions/py39_cu117/cpu_adam...
Detected CUDA files, patching ldflags
Emitting ninja build file /root/.cache/torch_extensions/py39_cu117/cpu_adam/build.ninja...
Building extension module cpu_adam...
Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
[1/3] /usr/local/cuda/bin/nvcc  -DTORCH_EXTENSION_NAME=cpu_adam -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1011\" -I/local_disk0/.ephemeral_nfs/envs/pythonEnv-615f6812-9c5d-4630-9be1-4a471fc90e4b/lib/python3.9/site-packages/deepspeed/ops/csrc/includes -I/usr/local/cuda/include -isystem /databricks/python/lib/python3.9/site-packages/torch/include -isystem /databricks/python/lib/python3.9/site-packages/torch/include/torch/csrc/api/include -isystem /databricks/python/lib/python3.9/site-packages/torch/include/TH -isystem /databricks/python/lib/python3.9/site-packages/torch/include/THC -isystem /usr/local/cuda/include -isystem /usr/include/python3.9 -D_GLIBCXX_USE_CXX11_ABI=0 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr -gencode=arch=compute_86,code=compute_86 -gencode=arch=compute_86,code=sm_86 --compiler-options '-fPIC' -O3 --use_fast_math -std=c++14 -U__CUDA_NO_HALF_OPERATORS__ -U__CUDA_NO_HALF_CONVERSIONS__ -U__CUDA_NO_HALF2_OPERATORS__ -gencode=arch=compute_86,code=sm_86 -gencode=arch=compute_86,code=compute_86 -c /local_disk0/.ephemeral_nfs/envs/pythonEnv-615f6812-9c5d-4630-9be1-4a471fc90e4b/lib/python3.9/site-packages/deepspeed/ops/csrc/common/custom_cuda_kernel.cu -o custom_cuda_kernel.cuda.o 
[2/3] c++ -MMD -MF cpu_adam.o.d -DTORCH_EXTENSION_NAME=cpu_adam -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1011\" -I/local_disk0/.ephemeral_nfs/envs/pythonEnv-615f6812-9c5d-4630-9be1-4a471fc90e4b/lib/python3.9/site-packages/deepspeed/ops/csrc/includes -I/usr/local/cuda/include -isystem /databricks/python/lib/python3.9/site-packages/torch/include -isystem /databricks/python/lib/python3.9/site-packages/torch/include/torch/csrc/api/include -isystem /databricks/python/lib/python3.9/site-packages/torch/include/TH -isystem /databricks/python/lib/python3.9/site-packages/torch/include/THC -isystem /usr/local/cuda/include -isystem /usr/include/python3.9 -D_GLIBCXX_USE_CXX11_ABI=0 -fPIC -std=c++14 -O3 -std=c++14 -g -Wno-reorder -L/usr/local/cuda/lib64 -lcudart -lcublas -g -march=native -fopenmp -D__AVX256__ -D__ENABLE_CUDA__ -c /local_disk0/.ephemeral_nfs/envs/pythonEnv-615f6812-9c5d-4630-9be1-4a471fc90e4b/lib/python3.9/site-packages/deepspeed/ops/csrc/adam/cpu_adam.cpp -o cpu_adam.o 
[3/3] c++ cpu_adam.o custom_cuda_kernel.cuda.o -shared -lcurand -L/databricks/python/lib/python3.9/site-packages/torch/lib -lc10 -lc10_cuda -ltorch_cpu -ltorch_cuda_cu -ltorch_cuda_cpp -ltorch -ltorch_python -L/usr/local/cuda/lib64 -lcudart -o cpu_adam.so
Loading extension module cpu_adam...
Time to load cpu_adam op: 31.890299081802368 seconds
Using /root/.cache/torch_extensions/py39_cu117 as PyTorch extensions root...
Creating extension directory /root/.cache/torch_extensions/py39_cu117/utils...
Emitting ninja build file /root/.cache/torch_extensions/py39_cu117/utils/build.ninja...
Building extension module utils...
Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
[1/2] c++ -MMD -MF flatten_unflatten.o.d -DTORCH_EXTENSION_NAME=utils -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1011\" -isystem /databricks/python/lib/python3.9/site-packages/torch/include -isystem /databricks/python/lib/python3.9/site-packages/torch/include/torch/csrc/api/include -isystem /databricks/python/lib/python3.9/site-packages/torch/include/TH -isystem /databricks/python/lib/python3.9/site-packages/torch/include/THC -isystem /usr/include/python3.9 -D_GLIBCXX_USE_CXX11_ABI=0 -fPIC -std=c++14 -c /local_disk0/.ephemeral_nfs/envs/pythonEnv-615f6812-9c5d-4630-9be1-4a471fc90e4b/lib/python3.9/site-packages/deepspeed/ops/csrc/utils/flatten_unflatten.cpp -o flatten_unflatten.o 
[2/2] c++ flatten_unflatten.o -shared -L/databricks/python/lib/python3.9/site-packages/torch/lib -lc10 -ltorch_cpu -ltorch -ltorch_python -o utils.so
Loading extension module utils...
Time to load utils op: 16.06422710418701 seconds
Parameter Offload: Total persistent parameters: 1070080 in 258 params
[2023-06-07 02:41:19,981] [INFO] [launch.py:318:sigkill_handler] Killing subprocess 4893
[2023-06-07 02:41:19,991] [ERROR] [launch.py:324:sigkill_handler] ['/local_disk0/.ephemeral_nfs/envs/pythonEnv-615f6812-9c5d-4630-9be1-4a471fc90e4b/bin/python', '-u', '-m', 'training.trainer', '--local_rank=0', '--input-model', 'EleutherAI/pythia-2.8b', '--deepspeed', '/Workspace/Repos/xxxxxx/dolly/config/a10_config.json', '--epochs', '2', '--local-output-dir', '/local_disk0/dolly_training/dolly__2023-06-07T02:38:06', '--dbfs-output-dir', '/dbfs/dolly_training/dolly__2023-06-07T02:38:06', '--per-device-train-batch-size', '4', '--per-device-eval-batch-size', '4', '--logging-steps', '10', '--save-steps', '200', '--save-total-limit', '20', '--eval-steps', '50', '--warmup-steps', '50', '--test-size', '200', '--lr', '5e-6', '--bf16', 'true'] exits with return code = -9

**Comment by srowen** [comment]
Not enough, no. In this case it's OS memory; I think 64GB is actually not enough. g5.8xlarge might work. That's just one GPU though, will take a while to train

**Comment by ipfloater** [comment]
Got it. Thank you for the clarification. Will try to get hold of **g5.8xlarge**.
